### DISCUSSION TITLE
Best way to covert to and from IGL (array of vertices coord and array of vertex indices for face)

### DISCUSSION THREAD

**Discussion by Nosenzor** [pmp-library/pmp-library#173] [General]
Hi all,
Many geometric libraries would rather use the array format
V=Array of coordinates + F=Array of triple indices into V (Triangles).

What would be the best way to create a pmp::SurfaceMesh from those two arrays ?
And what'd be the best way to export a pmp::SurfaceMesh to those two arrays ?

**Comment by dsieger** [comment]
Disclaimer: I never worked with IGL 😱, so I maybe I don't know what I'm doing!

Here's a function that creates a `pmp::SurfaceMesh` from IGL-style Eigen matrices:

```cpp
void igl_to_mesh(const Eigen::MatrixXd& V, const Eigen::MatrixXi& F, pmp::SurfaceMesh& mesh)
{
    for (int i = 0; i < V.rows(); i++)
    {
        auto x = static_cast<pmp::Scalar>(V(i, 0));
        auto y = static_cast<pmp::Scalar>(V(i, 1));
        auto z = static_cast<pmp::Scalar>(V(i, 2));
        pmp::Point p(x, y, z);
        mesh.add_vertex(p);
    }

    for (int i = 0; i < F.rows(); i++)
    {
        pmp::Vertex a(F(i, 0));
        pmp::Vertex b(F(i, 1));
        pmp::Vertex c(F(i, 2));
        mesh.add_triangle(a, b, c);
    }
}
```

And here's a function to convert a `pmp::SurfaceMesh` to IGL:

```cpp
void mesh_to_igl(const pmp::SurfaceMesh& mesh, Eigen::MatrixXd& V, Eigen::MatrixXi& F)
{
    V.resize(mesh.n_vertices(), 3);
    for (auto v : mesh.vertices())
    {
        auto p = mesh.position(v);
        auto i = v.idx();
        V(i, 0) = p[0];
        V(i, 1) = p[1];
        V(i, 2) = p[2];
    }

    F.resize(mesh.n_faces(), 3);
    for (auto f : mesh.faces())
    {
        int j{0};
        auto i = f.idx();
        for (auto v : mesh.vertices(f))
            F(i, j++) = v.idx();
    }
}
```

This is the most straightforward way I could think of. I'm quite sure there are more efficient ways.

I might actually add those functions to the library, I'd guess there are more folks interested in IGL interoperability.

Hope this helps!

**Reply by Nosenzor** [reply]
It would help. Thanks a million!

**Comment by dsieger** [comment]
I added two functions `matrices_to_mesh()` and `mesh_to_matrices()` in 720ed3d
